🐛 Fixed member chart flat growth for GMT -X Timezones.

iamscottcab opened the pull request:
closes #12083
Fixes a parsing issue where negative offset values were incorrectly having the `+` sign added regardless of actual offset for sqlite databases.

For mysql databases absolute values of offset were taken with sign applied where appropriate to stop issues where both hours and minutes could be negative which would cause both an issue with offsets that could present as `-2:-30` and by the look of the code also trigger extra padding to result in `-2:-030` rather than the expected `-2:30`. I'm not sure how to test this against a mysql db as it looks like sqlite comes out of the box for the dev environment. Any way I can double check the mysql fix locally? All I did was run a bunch of console logs against different timezone offsets to make sure they all only had one `-` at the front of hours and not minutes.

Got some code for us? Awesome 🎊!

Please include a description of your change & check your PR against this list, thanks!

- [x] There's a clear use-case for this code change
- [x] Commit message has a short title & references relevant issues
- [x] The build will pass (run `yarn test` and `yarn lint`)

More info can be found by clicking the "guidelines for contributing" link above.

Commit:
Fixed member chat flat growth for GMT -X Timezones.

refs #12083
Fixes a parsing issue where negative offset values were incorrectly having the + sign added regardless of actual offset for sqlite databases.

For mysql databases absolute values of offset were taken with sign applied where appropriate to stop issues where both hours and minutes could be negative which would cause both an issue with offsets that could present as -2:30 and by the look of the code also trigger extra padding to result in -2:-030 rather than the expected -2:30

allouis commented:
Hey! Thanks for the contribution!

It would be great if you could include some tests for this, just so we know that it _does_ work with MySQL :relaxed: 

If you need to spin up a local MySQL for testing you can use docker like:
```
docker run --name ghost-db -e MYSQL_ROOT_PASSWORD=password -p 3306:3306 -d mysql:5.7
```
and then add this to your `config.local.json` or `config.development.json` 
```json
{
    "database": {
        "client": "mysql",
        "connection": {
            "host": "127.0.0.1",
            "port": 3306,
            "user": "root",
            "password": "password",
            "database": "ghost"
       }
    }
}
```

After you've finished testing, if you want to clean up your system you can run `docker stop ghost-db`, `docker rm ghost-db` and `docker rmi mysql:5.7` to remove the associated container and image

iamscottcab commented:
@allouis cheers, no idea why I never thought to spin up a container. I'll try and get to this either this afternoon or in a few days when I have some spare time.

Thanks for the feedback :)

Commit:
Added test cases for GMT -/+ X timezones

Minor refactor of test suite to pull down member data and then test the return object against valid return data.

iamscottcab commented:
@allouis ignoring the fact that my test is failing on Node 14 (it's not locally on 12 and 10 (so I guess I will install Node 14 tomorrow and debug it) can you provide some general advice about the location and shape of the test in case you want me to refactor before I debug?

Basically I couldn't find a better home for this despite it feeling like a unit test and not an acceptance test. As I can reuse the same initial `supertest` request from `Can fetch stats` in all three tests I've had to do some lift and shift refactor so we can just reuse the same method a few times to prevent copy & paste. The problem I have is there are no other examples in that acceptance test file where you do that.

Would you prefer copy paste in each test or is there a better way I can handle that?

Commit:
Fixed test to run properly in UTC Timezone

endDate would have been inclusive in a UTC Timezone so just remove an hour to always make the endDate exclusive.

Commit:
fixed user import with no date breaking graph

Users imported from CSV with no created_at date where having their created_at date being stored as an int rather than a datetime. This was causing parsing issues with the graph. Fixed the formatting.

iamscottcab commented:
Finally it's all green....

daniellockyer commented:
@iamscottcab Thanks for updating the PR! I'm happy with your initial fix and it does seem to fix the graph.

However, I'm curious about your follow-up commit: [`fc3c78f` (#12168)](https://github.com/TryGhost/Ghost/pull/12168/commits/fc3c78f0d30c212735d88238f1ff2906be3ca810). I can't see where we'd end up storing an integer rather than a datetime. I tried importing a member using the code from `master` and it looks like the correct format in the DB, unless I'm missing something.

iamscottcab commented:
@daniellockyer sure I caught this when I was testing the `all-time` feature of the graph. I think this only effects **sqlite** databases as I'm just quickly doing a re-test now on MySQL and it seems fine but none-the-less it needs to be patched.

Looking at `core\server\services\members\importer\index.js` line `249-255` the variable `createdAt` is just a normal date. Basically just a `long` that represents the date. It gets parsed and inserted to the database as below in red (compared to the blue of a manual import).

![sqlite](https://user-images.githubusercontent.com/42525567/100025531-e6ca5c80-2e38-11eb-8016-96be6273e9fe.png)

So you're not explicitly storing them as a `long` but they are getting parsed that way when they get inserted. That causes the sql queries run in `core\server\api\canary\members.js` which populate the graph to return a `row` object with a `created_at` date of `null` rather than the desired date. This breaks the initial date on the graph and the number of members rendered until it gets to the next valid date in the response. By formatting the date when we run an input to match the same as when we manually create the user this fixes the graph. Additionally it doesn't seem to break other functionality that I smoke tested related to dates. Which makes sense because both the former and latter both parse to a valid date it's just the the sql query doesn't work when the dates are stored as a `long`.

I'm happy to discuss the semantics of that further or raise it as a different bug but it seemed relevant enough to the graph that I fix it as part of this PR.

daniellockyer commented:
@iamscottcab Thanks for the explanation! I can confirm I can reproduce that, and that the commit fixes it so I will merge that commit too 🙂

daniellockyer commented:
Bleugh, I rebased locally and merged but it seems it didn't get picked up here 😬 Anyway, it's merged in two commits:
* https://github.com/TryGhost/Ghost/commit/1e5d201835160f8ac6653c933ab60b14010157ea
* https://github.com/TryGhost/Ghost/commit/da5586e5aa75455462e2b098fd13a898297c49f6

daniellockyer commented:
Closing as we have now merged this as above 🙂

iamscottcab commented:
Phew I thought I'd done something terrible when I saw the red closed in my notifications. Thanks for being patient with that PR and guiding me through a few extra additions like tests, thanks for merging it in :)
